tpm2/marshalling.go

```diff
@@ -0,0 +1,237 @@
+func (m maybe[_]) OK() bool {
+	return m.err == nil
+}
+
+// Unwrap unwraps the result of an unmarshalling operation.
+// Panics if the data was not unmarshalled.
+func (m maybe[T]) Unwrap() *T {
+	if m.err != nil {
+		panic(fmt.Sprintf("could not unwrap: %v", m.err))
+	}
+	return m.value
+}
+
+// CheckUnwrap unwraps the result of an unmarshalling operation.
+func (m maybe[T]) CheckUnwrap() (*T, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	return m.value, nil
+}
+
+// asMaybe returns a maybe enclosing the given contents.
+func asMaybe[T any](t *T) maybe[T] {
+	return maybe[T]{
+		value: t,
+	}
+}
+
+// maybeNot returns a maybe with the given error.
+func maybeNot[T any](err error) maybe[T] {
+	return maybe[T]{
+		err: err,
+	}
+}
+
+// Marshallable represents any TPM type that can be marshalled.
+type Marshallable interface {
+	// marshal will serialize the given value, appending onto the given buffer.
+	// Returns an error if the value is not marshallable.
+	marshal(buf *bytes.Buffer)
+}
+
+// Unmarshallable represents any TPM type that can be marshalled or unmarshalled.
+type Unmarshallable interface {
+	Marshallable
+	// marshal will deserialize the given value from the given buffer.
+	// Returns an error if there was an unmarshalling error or if there was not
+	// enough data in the buffer.
+	unmarshal(buf *bytes.Buffer) error
+}
+
+// UnmarshallableWithHint represents any TPM type that can be marshalled or unmarshalled,
+// but that requires a selector ("hint") value when unmarshalling. Most TPMU_ are
+// an example of this.
+type UnmarshallableWithHint interface {
+	// create will instantiate and return the corresponding union member.
+	create(hint int64) (reflect.Value, error)
+	// get will return the corresponding union member by copy. If the union is
+	// uninitialized, it will initialize a new zero-valued one.
+	get(hint int64) (reflect.Value, error)
+}
+
+// Marshal will serialize the given values, returning them as a byte slice.
+func Marshal[T Marshallable](v T) []byte {
+	var buf bytes.Buffer
+	if err := marshal(&buf, reflect.ValueOf(v)); err != nil {
+		panic(fmt.Sprintf("unexpected error marshalling %v: %v", reflect.TypeOf(v).Name(), err))
+	}
+	return buf.Bytes()
+}
+
+// Unmarshal unmarshals the given type from the byte array.
+// Returns an error if the buffer does not contain enough data to satisfy the
+// types, or if the types are not unmarshallable.
+func Unmarshal[T any, P interface {
+	// *T must satisfy Marshallable
+	*T
+	Unmarshallable
+}](data []byte) maybe[T] {
+	buf := bytes.NewBuffer(data)
+	var t T
+	value := reflect.New(reflect.TypeOf(t))
+	if err := unmarshal(buf, value.Elem()); err != nil {
+		return maybe[T]{
+			err: err,
+		}
+	}
+	return maybe[T]{
+		value: value.Interface().(*T),
+	}
+}
+
+// marshallableByReflection is a placeholder interface, to hint to the unmarshalling
+// library that it is supposed to use reflection.
+type marshallableByReflection interface {
```

Comment:
Try making this more polymorphic with Marshallable

Reply:
Oh, the reason this didn't work out was that the marshalByReflection ends up as the receiver type, and you can't use reflect to deal with the other stuff